doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,196 @@
+
+A capability presents the `id` and `frontier` values, but is identified by a system-assigned unique identifier.
+Any user presenting this identifier has the associated capability to read from or write to the associated pTVC.
+Actions which read or write will validate the capability (e.g. in case a lease has expired) and may return errors if the capability is not valid as stated.
+
+Capabilities are durable, and their state is tracked by the system who associates a unique identifier with each.
+Capabilities can be cloned and transfered to others, providing the same rights (and responsibilities).
+Capabilities can be downgraded to frontiers that are greater or equal to the held frontier, which may then allow the `since` or `upper` frontier to advance.
+Capabilities can be dropped, which releases the constraint on the frontier, and the whichever ability (read or write) the capability provided.
+These actions are mediated by the system, and take effect only when the system confirms the action.
+Users can defer and batch these actions, modulo lease timeouts, imposing only on the freshness and efficiency of the system.
+
+Capabilities may expire, as a lease, which may result in errors from attempts to exercise the capability.
+
+It is very common for users to immediately drop capabilities if they do not require the ability.
+The write capability for most TVCs is dropped immediately (by the system) unless it is meant to be written to by users (e.g. a `TABLE`).
+The read capability may be dropped immediately if the user does not presently need to read the data.
+Capabilities may be recovered by the best-effort command `AcquireCapabilities(id)` to be discussed.
+
+## Storage
+
+The Storage layer presents a narrow API about which it makes guarantees.
+It likely has a more verbose diagnostic API that describes its state, which should be a view over the results of these commands.
+
+*   `Create(id, description) -> (ReadCapability, WriteCapability)`: binds to `id` the TVC described by `description`.
+
+    The command returns capabilities naming `id`, with frontiers set to initial values chosen by the Storage layer.
+
+    The collection associated with `id` is based on `description`, but it is up to the Storage layer to determine what the source's contents will be.
+    A standarad example is a Kafka topic, whose contents are "added" to the underlying collection as they are observed.
+    The Storage layer only needs to ensure that it can reliably produce the updates between `since` and `upper`, and otherwise the nature of the content is up for discussion.
+    It is an error to re-use a previously used `id`.
+
+*   `SubscribeAt(ReadCapability(id, frontier))`: returns a snapshot of `id` at `frontier`, followed by an ongoing stream of subsequent updates.
+
+    This command returns the contents of `id` as of `frontier` once they are known, and updates thereafter once they are known.
```

Comment:
If this ReadCapability's lease expires while the SubscribeAt is being used, how does that failure present itself?

Or, more concretely, some dataflowd is being a good citizen and uses a lease with duration X. It falls behind by more than X (for whatever reason you like), what happens?

(Or, maybe, does this question fundamentally misunderstand what the duration of a lease is represents?)

Reply:
In my mind, a subscription can blow up at any point. Your ability to restart the subscription depends on holding a read capability. The capability provides the introduction to the subscription, rather than being the beating heart of the subscription. If STORAGE is unable to continue to service a subscription (the outbound buffer is full and a day behind) it blows up the subscription, and when the consumer attempts to exercise their read capability to restart it, the systems says "sorry, that's not valid anymore; please call `AcquireCapabilities`".